Issue #6722: new check MissingLeadingAsteriskCheck

Gaurav-Punjabi opened the pull request:
Fixes #6722 

Contrib repo PR : [checkstyle/contribution#467](https://github.com/checkstyle/contribution/pull/467)
Sevntu checkstyle repo PR : [sevntu-checkstyle/sevntu.checkstyle#812](https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/812)
Diff Regression Report: https://gaurav-punjabi.github.io/checkstyle/6722/index.html
Fixes for violations: [checkstyle/checkstyle#8111](https://github.com/checkstyle/checkstyle/pull/8111)

Gaurav-Punjabi commented:
@strkkk I have added the check as requested in the issue. But this check is creating violations in our existing code having HTML content as show. One of the examples of such comment is shown below:
```
    /**
     * Represents the policy that static imports are all at the top.
     * For example:
     *
     * <pre>
        import static java.awt.Button.ABORT;
        import static java.io.File.createTempFile;
        import static javax.swing.WindowConstants.*;

        import java.awt.Button;
        import java.awt.event.ActionEvent;
     * </pre>
     */
```
Please suggest what should be done to resolve these. 
Should we ignore the leading asterisk for HTML content? 
Should Create a property for ignoring asterisk for HTML content 
OR
Fix all the violations?

Gaurav-Punjabi commented:
@strkkk Also I have updated the docs for this. But I'm a little confused with `Example of Usages` subsection. Please have a look at it and let me know if it's correct or not.

strkkk commented:
@Gaurav-Punjabi I suggest to fix them in second commit/separate PR.

Also, your check has many items to correct. Please see this PR as reference - https://github.com/checkstyle/checkstyle/pull/7935

Gaurav-Punjabi commented:
> @Gaurav-Punjabi I suggest to fix them in second commit/separate PR.

Okay 👍 

> Also, your check has many items to correct. Please see this PR as reference - #7935

I'll look into this PR and make the necessary changes.

Gaurav-Punjabi commented:
@strkkk I have fixed all the violations and added a new commit for that. Also, I have modified my check with reference to the PR you provided. Please have a look and let me know if anything other thing is wrong. I'll fix it.

strkkk commented:
@Gaurav-Punjabi you need to make CI green. 
travis says you have spelling errors, wercker says that there is violations in sevntu-checkstyle
```
[ERROR] src/main/java/com/github/sevntu/checkstyle/checks/annotation/RequiredParameterForAnnotationCheck.java:[31,1] (javadoc) MissingLeadingAsterisk: Missing leading asterisk
```

Please fix both (PR to sevntu is required to fix the latter one)

Gaurav-Punjabi commented:
> @Gaurav-Punjabi you need to make CI green.
> travis says you have spelling errors, wercker says that there is violations in sevntu-checkstyle
> 
> ```
> [ERROR] src/main/java/com/github/sevntu/checkstyle/checks/annotation/RequiredParameterForAnnotationCheck.java:[31,1] (javadoc) MissingLeadingAsterisk: Missing leading asterisk
> ```
> 
> Please fix both (PR to sevntu is required to fix the latter one)
> 
> And yes, diff report is necessary

Okay, I'll ping you once this is done.

Gaurav-Punjabi commented:
@strkkk Fixed the issues, added a link to the regression report and PR to seventu in the description of this PR.
Please review.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+ * </pre>
+ *
+ * @since 6.0
+ *
```
extra line

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+ * public void test() {}
+ * </pre>
+ *
+ * @since 6.0
```
version should be 8.32 (it always should be next version)

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+ *  * Another line of description // OK
+ *    Another line of description // violation, missing leading asterisk at the start
+ *  *&#47;
+ * public void test() {}
```
add 
1. surrounding class
2. Single like javadoc example

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+     * @return true, if the given comment is an inline comment.
+     */
+    private static boolean isInlineComment(String comment) {
+        return comment.split(COMMENT_SEPARATOR).length == 1;
```
`comment.split(COMMENT_SEPARATOR).length == 1;` it is better to look for next BLOCK_COMMENT_END token

Gaurav-Punjabi replied:
Sure

Gaurav-Punjabi replied:
So to check if a comment is inline, I will check if there are any new lines between `BLOCK_COMMENT_START` and `BLOCK_COMMENT_END` correct?

strkkk replied:
I guess you meant that comment is **not** inline.
Yes.

Gaurav-Punjabi replied:
Done, updated this method.

Gaurav-Punjabi replied:
@strkkk I removed this method because we don't need it. As there are no newlines in an inline document no violations can occur from it. So it doesn't matter whether a comment is inline or not.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+     * @param comment content of the comment.
+     * @return true, if the given comment is an javadoc comment.
+     */
+    private static boolean isJavaDocComment(String comment) {
```
There is existing `JavadocUtil.isJavadocComment`, please do not reinvent it :)

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/messages.properties:
```diff
@@ -13,6 +13,7 @@ javadoc.invalidInheritDoc=Invalid use of the '{'@inheritDoc'}' tag.
 javadoc.legacyPackageHtml=Legacy package.html file should be removed.
 javadoc.missed.html.close=Javadoc comment at column {0} has parse error. Missed HTML close tag ''{1}''. Sometimes it means that close tag missed for one of previous tags.
 javadoc.missing=Missing a Javadoc comment.
+javadoc.missing.asterisk=Missing leading asterisk
```
As for me, `Javadoc line is missing leading asterisk.` sounds better.
Please also mind the dot at the end.

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyForInitializerPadCheckTest.java:
```diff
@@ -91,7 +92,8 @@ public void testPadOptionValueOf() {
 
-    /* Additional test for jacoco, since valueOf()
+    /**
+     * Additional test for jacoco, since valueOf()
```
how this changes are related with this check?

Gaurav-Punjabi replied:
Initially, this comment was creating a violation, so to fix it I made this change. Then later to improve the check I added a method `isJavadocComment` to only validate the comments if they are Javadoc comments. Hence now this won't create a violation. I'll revert this change.

Gaurav-Punjabi replied:
Done

strkkk replied:
these changes are still here. I suggest to create different PR for corrections for this check to avoid confusion and leave here only changes necessary for new check

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,20 @@
+  /*
+      This is a non javadoc comment.
+   */
+  /**/ // This is empty comment for 100% code coverage.
```
`// This is empty comment for 100% code coverage.` please remove this comment and add one more example - empty single line javadoc `/***/`

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsteriskFalsePositive.java:
```diff
@@ -0,0 +1,9 @@
+package com.puppycrawl.tools.checkstyle.checks.javadoc.missingleadingasterisk;
+
+public class InputMissingLeadingAsteriskFalsePositive {
```
why it is false positive?

Gaurav-Punjabi replied:
My bad. This is a positive test case. I'll change the name.

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,20 @@
+package com.puppycrawl.tools.checkstyle.checks.javadoc.missingleadingasterisk;
+
+public class InputMissingLeadingAsterisk {
```
Please also add config description, like
```
/** Config = default */
```

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,20 @@
+       violation
+   violation
+
+   **/
```
please add examples with tag and param on line start

Gaurav-Punjabi replied:
Done

strkkk commented on src/xdocs/config_javadoc.xml:
```diff
@@ -2554,6 +2554,68 @@ class DatabaseConfiguration {}
     </section>
 
+    <section name="MissingLeadingAsterisk">
+      <p>Since Checkstyle 6.0</p>
```
here version should be 8.32

Gaurav-Punjabi replied:
Done

strkkk commented:
@Gaurav-Punjabi 
Also in report there are some weird cases like 
https://gaurav-punjabi.github.io/checkstyle/issue-6722/hibernate-orm-with-excludes/xref/Users/gauravpunjabi/checkstyle-contribution/checkstyle-tester/repositories/hibernate-orm-with-excludes/hibernate-core/src/main/java/org/hibernate/engine/spi/PersistenceContext.java.html#L600

They should be reflected in UT inputs

Gaurav-Punjabi commented:
> They should be reflected in UT inputs

@strkkk Added these cases in the UT.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,76 @@
+package com.puppycrawl.tools.checkstyle.checks.javadoc.missingleadingasterisk;
+
+/** Config = default */
+public class InputMissingLeadingAsterisk {
```
please add `// violation` comment to each line where violation is reported

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,76 @@
+      This is a non javadoc comment.
+   */
+  /**/
+  /***/
```
this is a dangling javadoc, not very good. Please add some method/field after it.

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsteriskCorrect.java:
```diff
@@ -0,0 +1,9 @@
+package com.puppycrawl.tools.checkstyle.checks.javadoc.missingleadingasterisk;
+
+public class InputMissingLeadingAsteriskCorrect {
```
No need for separate test case and input. Other test input also has correct examples.
Please remove it with test

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,117 @@
+ * Checks if the javadoc has leading asterisk on each line or not.
+ * </p>
+ * <p>
+ * To configure the default check:
```
`default` there is no other options than default, so this word is not needed here

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,117 @@
+ * <pre>
+ * &#47;**
+ *  * Some description here. // OK
+ *    Another line of description. // violation, missing leading asterisk at the start
```
actually, violation is reported on javadoc start, so violation comment should be on other line

Gaurav-Punjabi replied:
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,117 @@
+ *   &#47;**
+ *    * Some description here. // OK
+ *    * Another line of description. // OK
+ *      Another line of description. // violation, missing leading asterisk at the start
```
same here

Gaurav-Punjabi replied:
Done

Gaurav-Punjabi commented:
I've updated the diff regression report too.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,81 @@
+  /**
+   * Some description here.
+   * <p>
+     Some HTML content here.
```
should be a violation comment

Gaurav-Punjabi replied:
Done.

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,81 @@
+
+  /**
+   * Some description here.
+   << * Another line of description.
```
same here

Gaurav-Punjabi replied:
Done

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,81 @@
+  public void test2() {}
+
+  /*
+      This is a non javadoc comment.
```
same here

Gaurav-Punjabi replied:
This line does not create a violation as it isn't a Javadoc.

strkkk replied:
sorry

strkkk commented:
@Gaurav-Punjabi please also move violation fixes in separate PR

Gaurav-Punjabi commented:
> @Gaurav-Punjabi please also move violation fixes in separate PR

Okay

Gaurav-Punjabi commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * The new line character that separates each line of javadoc.
+     */
+    private static final String COMMENT_SEPARATOR = "\n";
```
I can iterate through each `AST` child instead until I find `BLOCK_COMMMENT_END`. Is this okay?

Gaurav-Punjabi replied:
@strkkk I just checked it out. We can't iterate through the AST nodes. Because the whole comment content is one AST node. Can you suggest me some other way?

strkkk replied:
You can take line number of block begin and block end, after it get file content https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/api/FileContents.java
and iterate over lines there.
Usage of this class is not encouraged, but as for me it looks better than possible problems with line separators

Gaurav-Punjabi replied:
I tried this method. But the `getLine()` method in `FileContents` returns the following String
` * // violation\n  ** OK\n *\n  *\n violation`. It is wrapping all the comment lines into one single string.

strkkk replied:
Ok. But this line separator can be a problem. Can you try with CR line separator? I am afraid it wont report violation in this case.

Gaurav-Punjabi replied:
Sure. Let me add more cases to the input.

Gaurav-Punjabi replied:
@strkkk I have an idea. The different line separators LN, CR, CRLN are represented as `\n`, `\r`, `\n\r` in `String`. So what if we write a Regex and pass this regex to `split` function of `String`.

strkkk replied:
We can try this

Gaurav-Punjabi replied:
I have updated the code. Added the regex here https://github.com/checkstyle/checkstyle/pull/8085/files#diff-1c1129c743f863d78fcd7fba8d0a81e9R73
Please have a look

Gaurav-Punjabi commented on src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionMissingLeadingAsteriskTest.java:
```diff
@@ -0,0 +1,61 @@
+                createModuleConfig(MissingLeadingAsteriskCheck.class);
+
+        final String[] expectedViolation = {
+            "3:1: " + getCheckMessage(JavadocStyleCheck.class,
```
It's a mistake. I'll fix it.

Gaurav-Punjabi replied:
Done.

strkkk replied:
I still see it.

Gaurav-Punjabi replied:
Fixed it 👍

Gaurav-Punjabi commented:
> @Gaurav-Punjabi please also move violation fixes in separate PR

Done. Added a link to the PR in the description

Gaurav-Punjabi commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,108 @@
+ *    *&#47;
+ *   public void test() {}
+ *
+ *   &#47;** Some description here. *&#47; // OK
```
Done

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,116 @@
+        final String commentContent = JavadocUtil.getBlockCommentContent(ast);
+        if (JavadocUtil.isJavadocComment(commentContent)) {
+            for (String singleLine : MATCH_COMMENT_SEPARATOR.split(commentContent)) {
+                if (!CommonUtil.isBlank(singleLine) && singleLine.trim().charAt(0) != '*') {
```
I have just noticed that we skip empty lines inside javadoc.
Why do we do this?
Cases like
```
/**
 * some text

 * some other text
 */
```
look like a violation to me

Gaurav-Punjabi replied:
Oh yes, I missed that.

Gaurav-Punjabi replied:
Done. Also updated the newly generated violations due to this in PR #8111

strkkk commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/missingleadingasterisk/InputMissingLeadingAsterisk.java:
```diff
@@ -0,0 +1,81 @@
+  public void testWithEmptyComment() {}
+
+  /***/
+  public void testWithEmptyJavadocComment() {}
```
please also add example for case with just empty line
```
/**
 * some text

 * some other text
 */
```

Gaurav-Punjabi replied:
Done.

strkkk commented:
Please also rebase to latest master to fix travis build

Gaurav-Punjabi commented:
> Please also rebase to latest master

Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+            // We need to ignore the last string.
+            // As a java doc always has a empty string at the end, which is not a violation.
+            for (int i = 0; i < commentLines.length - 1; i++) {
+                final String comment = commentLines[i];
```
`String comment` please rename to `commentLine`

Gaurav-Punjabi replied:
Done.

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+            // As a java doc always has a empty string at the end, which is not a violation.
+            for (int i = 0; i < commentLines.length - 1; i++) {
+                final String comment = commentLines[i];
+                if (CommonUtil.isBlank(comment) || comment.trim().charAt(0) != '*') {
```
`comment.trim().charAt(0) != '*'`
as for me `!comment.trim().startsWith("*")` looks better

Gaurav-Punjabi replied:
Done

Gaurav-Punjabi replied:
@strkkk IDEA inspection is recommending to replace `startsWith` with `charAt` as we are checking a single character.

strkkk replied:
Interesting. Ok, lets keep it with char comparison

Gaurav-Punjabi replied:
Yeah. Updated the code 👍

strkkk commented:
@Gaurav-Punjabi Also please
1. fix CI issues
2. Create new diff report since logic has changed (keep the old one to compare them)

Gaurav-Punjabi commented:
> @Gaurav-Punjabi Also please
> 
> 1. fix CI issues
> 2. Create new diff report since logic has changed (keep the old one to compare them)

Alright 👍

strkkk commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+            // As a java doc always has a empty string at the end, which is not a violation.
+            for (int i = 0; i < commentLines.length - 1; i++) {
+                final String commentLine = commentLines[i];
+                if (CommonUtil.isBlank(commentLine) || commentLine.trim().startsWith("*")) {
```
`commentLine.trim().startsWith("*")`
you forgot to negate this condition

Gaurav-Punjabi commented:
> 2. Create new diff report since logic has changed (keep the old one to compare them)

Updated the diff report. Link is in the description.

strkkk commented:
Diff report is empty, there is no new check in config.

Gaurav-Punjabi commented:
> Diff report is empty, there is no new check in config.

Please check again, updated the link. There were some issues with Github pages.

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,127 @@
+        if (JavadocUtil.isJavadocComment(commentContent)) {
+            final String[] commentLines = MATCH_COMMENT_SEPARATOR.split(commentContent);
+            // We need to ignore the last string.
+            // As a java doc always has a empty string at the end, which is not a violation.
```
Also, one-line Javadoc comments will be ignored here.
```
/** * Javadoc with a leading asterisk on the first line is legal, but weird. */
```
But what about [Scala-style](https://docs.scala-lang.org/style/scaladoc.html) (compact) comments:
```
/** First line
  * Second line
  * Third line
  */
```
?
And this one:
```
/** First line
  * Second line
    Third line  **/ // violation
```
?
 Please add them as test cases.

Gaurav-Punjabi replied:
> ```
> /** First line
>   * Second line
>   * Third line
>   */
> ```


This is okay, it will not generate any violations as this is a valid comment.

> ```
> /** First line
>   * Second line
>     Third line  **/ // violation
> ```

Yes, this one will generate a violation. I'll add them to the test case input.

> Also, one-line Javadoc comments will be ignored here.

We ignored inline Javadoc comments, you can see the details in this [conversation](https://github.com/checkstyle/checkstyle/pull/8085#discussion_r408239265)

Gaurav-Punjabi replied:
Added the given cases to UT input.

pbludov replied:
> Added the given cases to UT input.

I still see no cases like [this one](http://hg.openjdk.java.net/jdk10/master/file/be620a591379/src/java.base/share/classes/jdk/internal/reflect/ClassFileAssembler.java#l584)

Gaurav-Punjabi replied:
I added this test case to the input. But it didn't generate a violation, due to this code
```
// We need to ignore the last string.
// As a java doc always has an empty string at the end, which is not a violation.
 for (int i = 0; i < commentLines.length - 1; i++) {
```
To fix this, I was thinking to check the last string only if `BLOCK_COMMENT_END` and that comment are on the same line. What do you think? Is this okay?

pbludov replied:
Yes, this will work.

Gaurav-Punjabi replied:
There is just one problem. The whole comment is a single ASTNode. So I cannot use `TokenUtil.areOnSameLine`. I'll need to manually compare the line numbers.

pbludov replied:
@Gaurav-Punjabi, is it possible to ignore the last line if it contains only whitespaces?

Gaurav-Punjabi replied:
Let me check.

Gaurav-Punjabi replied:
@pbludov I don't think to verify the last line with a whitespace would be reliable. I'm saying this because not all the Javadoc end with a white space. Some end with a blank string or new line. 
##### Example 1:
```
/** // violation
   ** OK
   *
       violation
   violation

   **/
```
When this example is parsed it ends with a blank string.

##### Example 2:
Where as in the example below:
```
  /**
   * Some description here.
   * @param a description of the parameter.
   */
```
The last line is a single space.

Gaurav-Punjabi replied:
I would recommend to just compare lineNo of last line and `BLOCK_COMMENT_END` after the for loop ends. If they are on the same line, we will further validate it otherwise we'll just ignore it. I think this will work in all the case without any issues.

Gaurav-Punjabi replied:
@pbludov It worked! I have updated the code and added the given case to the UT Input. Please review, I'll provide a new regression report for this as the code has been updated.

Gaurav-Punjabi commented:
@pbludov PING

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,133 @@
+            // As a java doc always has a empty string at the end, which is not a violation.
+            for (int i = 0; i < commentLines.length - 1; i++) {
+                final String commentLine = commentLines[i];
+                if (CommonUtil.isBlank(commentLine) || commentLine.trim().charAt(0) != '*') {
```
Can this be just `!commentLine.trim().startsWith("*")`?

pbludov replied:
I think it is better to have one loop for all lines.
The condition will be "the trimmed line should start with "*" OR (be empty if last)

Gaurav-Punjabi replied:
> Can this be just `!commentLine.trim().startsWith("*")`?

@strkkk recommended the same thing. But IDEA Inspection was throwing an error that I should replace `startsWith` with `charAt` as we were comparing a single character. So we decided to use `charAt` instead.

Gaurav-Punjabi replied:
> I think it is better to have one loop for all lines.
> The condition will be "the trimmed line should start with "*" OR (be empty if last)

Done

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,128 @@
+        final String commentContent = JavadocUtil.getBlockCommentContent(ast);
+        if (JavadocUtil.isJavadocComment(commentContent)) {
+            final String[] commentLines = MATCH_COMMENT_SEPARATOR.split(commentContent);
+            // We need to ignore the last string.
```
Please remove this comment

Gaurav-Punjabi replied:
Done

pbludov commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,128 @@
+            // As a java doc always has a empty string at the end, which is not a violation.
+            for (int i = 0; i < commentLines.length; i++) {
+                final String commentLine = commentLines[i];
+                if (i + 1 != commentLines.length && CommonUtil.isBlank(commentLine)
```
Please extract this `if ` to a small method like
`boolean isValidLine(String line, boolean isLast)`
It is hard to read the existent code.

Gaurav-Punjabi replied:
Done

pbludov commented:
@Gaurav-Punjabi please do `git pull --rebase main master`. Your PR is based on a too old master commit. It is good to alvays rebase before pushing.

Gaurav-Punjabi commented:
> @Gaurav-Punjabi please do `git pull --rebase main master`

Done

>  It is good to alvays rebase before pushing.

Thanks, will keep that in mind next time 👍

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,139 @@
+
+/**
+ * <p>
+ * Checks if the javadoc has leading asterisk on each line or not.
```
remove `or not`.

make it 
`Checks that each line of javadoc comment has leading asterisk`

please try put link https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks in `leading asterisk`. We need to educate users on what is it.

Gaurav-Punjabi replied:
Done

romani commented on src/xdocs/config_javadoc.xml:
```diff
@@ -2691,6 +2691,90 @@ class DatabaseConfiguration {}
     </section>
 
+    <section name="MissingLeadingAsterisk">
+      <p>Since Checkstyle 8.32</p>
```
8.33

Gaurav-Punjabi replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,139 @@
+                final String commentLine = commentLines[i];
+                final boolean isLast = i + 1 == commentLines.length;
+                if (!isValidLine(commentLine, isLast)) {
+                    log(ast, MSG_MISSING_ASTERISK);
```
can we do ` for (int i = 0; i < commentLines.length - 1; i++) {` ?
and validate last line is special way after FOR-loop, logic in `isValidLine` is very complicated.

Gaurav-Punjabi replied:
Sure, we can iterate till `commentLines.length - 1` and the check the last line after the for loop ends. Is this okay ?

Gaurav-Punjabi replied:
Done

romani commented on src/xdocs/config_javadoc.xml:
```diff
@@ -2691,6 +2691,91 @@ class DatabaseConfiguration {}
+      <p>Since Checkstyle 8.33</p>
+      <subsection name="Description" id="MissingLeadingAsterisk_Description">
+        <p>
+          Checks if the javadoc has <a href="https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks">leading asterisk</a>
```
wrap long line.

Gaurav-Punjabi replied:
Done

romani commented on src/xdocs/checks.xml:
```diff
@@ -531,6 +531,13 @@
+          <tr>
+            <td><a href="config_javadoc.html#MissingLeadingAsterisk">MissingLeadingAsterisk</a></td>
+            <td>
+              Checks if the javadoc has <a href="https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks">leading asterisk</a>
```
wrap long line
only URL content is allowed to be non-breakable and long.

Gaurav-Punjabi replied:
Done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,145 @@
+            // To validate the last line of javadoc
+            final String lastCommentLine = commentLines[commentLines.length - 1];
+            if (!CommonUtil.isBlank(lastCommentLine) && lastCommentLine.trim().charAt(0) != '*') {
+                log(ast, MSG_MISSING_ASTERISK);
```
please show me what UT and line in Input covers this case.

> commentLines[commentLines.length - 1];

This might be a problem on empty comment, please also share link to Input that covers it, and explain why it is not failing.

Gaurav-Punjabi replied:
> please show me what UT and line in Input cover this case.

Here's [link](), please show me what UT and line in Input cover this case. 

> `commentLines[commentLines.length - 1];`
This might be a problem on empty comment, please also share link to Input that covers it, and explain why it is not failing.

Here's the [link]() to an empty comment input to the UT. As you can see in the input it has the following empty comment:
```
/***/
```
While parsing we get `"*"` as the content of Javadoc. The below statement then split's this string by new lines:
```
final String[] commentLines = MATCH_COMMENT_SEPARATOR.split(commentContent);
```
As there are no new lines present in the content, it returns `["*"]` as the result. The length of array is one due to which the for loop is skipped and finally the below condition, verifies that the string is not blank and it starts with a `*`:
```
if (!CommonUtil.isBlank(lastCommentLine) && lastCommentLine.trim().charAt(0) != '*') {
```
which is true, the string contains only `*`.

Hence, our code is not failing for empty comments.

Gaurav-Punjabi commented:
> please regenerate diff report. please take latest code from contribution repo.

Sure, I'll update the report and ping you once it's done

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingLeadingAsteriskCheck.java:
```diff
@@ -0,0 +1,145 @@
+                }
+            }
+
+            // To validate the last line of javadoc
```
please make it `// last line of javadoc could be blank`

Gaurav-Punjabi replied:
Done.

Gaurav-Punjabi commented:
@romani Updated the diff report. Couldn't generate the report for JDK 8/9/11. It is throwing this error while cloning `stream ended unexpectedly`.

romani commented:
>  Couldn't generate the report for JDK 8/9/11. It is throwing this error while cloning stream ended unexpectedly.

please share whole error context, whole log. you can activate only one jdk project and run checketyl-tester on it only, share whole log in gist.github.com

romani commented:
we have problem:
https://gaurav-punjabi.github.io/checkstyle/6722/RxJava/index.html#A4
https://gaurav-punjabi.github.io/checkstyle/6722/pmd/index.html#A2

embedded code:
```
* <pre> {@code
 
// observer will receive no onNext
AsyncSubject<Object> subject = AsyncSubject.create();
subject.subscribe(observer);
subject.onNext("one");
} </pre>
* 
* @param par
```

from https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks
>  If you omit the leading asterisk on a line, then the leading white space is no longer removed so that you can paste code examples directly into a documentation comment inside a <pre> tag with its indentation preserved. 

so this is valid use case. What we can do to avoid violations on such javadocs?

I need report on openjdk .... as there might be more examples of use usecase.

Gaurav-Punjabi commented:
> so this is a valid use case. What we can do to avoid violations on such javadocs?

We can check for HTML tag in the comment line. If they are found we'll just ignore the check rest lines until a closing HTML tag is found for the same. But this will create an issue if there is not closing HTML tag present in the doc

> I need report on OpenJDK .... as there might be more examples of use use case.

Yeah it's building, I'll share the logs soon

Gaurav-Punjabi commented:
@romani I have another idea for this problem. We can use `String` class `replaceAll` function to remove all the HTML content, so it won't be checked and will satisfy all the cases.

##### Example:
```
content.replaceAll("\\<\w*\\>.*\\</w*\\>", "");
```

It would save us a lot of string matching and will maintain the performance. But I'm not sure whether it will cover all the cases or not, it's just a rough idea. Will need to test and improve it.

What do you say ?

Gaurav-Punjabi commented:
@romani Here's the [link](https://gist.github.com/Gaurav-Punjabi/ea44b62897cd1edab18ea1a30a3df5f2) to the error log, generated while running diff-report-tool for JDK 10.
I'm able to generate the report for 7 and 8.

romani commented:
@Gaurav-Punjabi , please investigate a reason why RxJava refactored their code from missed leading asterik to javadoc with leading asterik. Missing leading asterik might be problematic practice, but we need to get to know it to explain it in our documentation and encourage users of this Check to add asterisks. Please investigate benefit of missed asterisk in such cases (javadoc tool might do somethig, or IDEs during rendering).
One obvious reason to skip asterisk as it is easy to copy-paste code in/out of javadoc, but with such idea we can skip asterisk at all, but people do not do this.

One simple workaround is: when we detect line without asterisk, we need to check previous line and it contains string `<pre>`(case insensitive) we skip violation and skip validating further till we get to next line with asterisk.

Gaurav-Punjabi commented:
@romani Alright, I'll do a little research and will let you know 👍 

> One simple workaround is: when we detect line without asterisk, we need to check previous line and it contains string `<pre>` (case insensitive) we skip violation and skip validating further till we get to next line with asterisk.

Sounds good.

I'll first investigate a little more about missing leading asterisk, then we'll discuss and implement the solution for these cases

romani commented:
@Gaurav-Punjabi ,

from your log:
> 
```
Testing Checkstyle started
   [delete] Deleting directory /Users/gauravpunjabi/checkstyle-contribution/
checkstyle-tester/src/main/java
Cloning hg repository 'openjdk10' to repositories/openjdk10 folder ...
transaction abort!
```

you have some problem with `hg`, I hope you have it installed, but it fails to download sources, please update https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on.properties to `openjdk14|git|https://github.com/openjdk/jdk14|master||`

Gaurav-Punjabi commented:
> please update https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on.properties to `openjdk14|git|https://github.com/openjdk/jdk14|master||`

@romani Should I update it locally or do you want me to update it on checkstyle/contribution repo ?

romani commented:
I already sent PR for contribution repo, please do it locally for now as PR might take a time to be merged.

Gaurav-Punjabi commented:
Got it 👍 
Thanks :)

Gaurav-Punjabi commented:
@romani There's no specific reason why RxJava decided to use a leading asterisk in their Javadoc. They still have missing leading asterisk in their branch 1.x as we can see [here](https://github.com/ReactiveX/RxJava/blob/1.x/src/main/java/rx/subjects/AsyncSubject.java#L39).
After [1.x](https://github.com/ReactiveX/RxJava/tree/1.x) in branch [2.x](https://github.com/ReactiveX/RxJava/tree/2.x) and [3.x](https://github.com/ReactiveX/RxJava) they rewrote `AsyncSubject` completely and while writing the docs for that class they must have added a leading asterisk to all the lines as seen [here](https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/subjects/AsyncSubject.java)

Also, I couldn't find any other reasons to miss leading asterisk other than to make copy/pasting easier. Java discards the leading asterisk, they are just used for styling and to make the Javadoc look good and clean.

romani commented:
ok, lets make reports for openjdk7 and openjdk14. We might see some pattern there.
Please add in inputs cases with `PRE` please start implementation to support this style, we need to make ability to allow now leading asterisk for PRE tag by special property(no violation by default) so by mean of this we will reconcile people expectations of validation and best-practice.

Gaurav-Punjabi commented:
> Please add in inputs cases with PRE please start implementation to support this style. 

Okay

___

> We need to make the ability to allow now leading asterisk for PRE tag(there might be other no-format HTML tags) by special property(no violation by default) so by mean of this we will reconcile people expectations of validation and best-practice.

We'll add a property named 'checkHTML` which will be true by default.

____

> can we write javadoc inner code with @code ?

Yes, we can. Here's an example for that:
```
* <pre>
* {@code
* Set<String> s;
* System.out.println(s);
* }
* </pre>
```

____


@romani openjdk10 and openjdk14 have been successfully cloned, but while generating the site `maven-site-plugin` is throwing a `NullPointerException`. I'll provide the logs for your reference.
The issue is similar to this comment https://github.com/checkstyle/checkstyle/pull/8093#issuecomment-617277511

romani commented:
> Here's an example for that:

please do same example without leading asterisk and show image of how javadoc tool render both cases.

Gaurav-Punjabi commented:
> please do same example without leading asterisk and show image of how javadoc tool render both cases.

I should add these examples in the xdocs, right ?

romani commented:
> I should add these examples in the xdocs, right ?

yes, we need to explain user all details, to predict their questions.

Gaurav-Punjabi commented:
@romani
1. Should we allow missed leading asterisk in inline pre/code tags as shown below? I don't think we should. As the main motto of allowing missed leading asterisk in code markups was to allow easy copy/pasting. But in inline tags, it won't matter.
```
/**
 * First line.
   <pre>Inline code markups</pre>
 */
```

romani commented:
this should be a violation, please keep this case in Inputs.

romani commented:
@Gaurav-Punjabi , please finish this PR.

romani commented:
@Gaurav-Punjabi , please finish this PR.

Gaurav-Punjabi commented:
> @Gaurav-Punjabi , please finish this PR.

Working on it 👍

romani commented:
@Gaurav-Punjabi , please let us know if there is some difficulties, we might advice something.

romani commented:
@Gaurav-Punjabi , ping.

romani commented:
@pbludov , do you have time to finish this implementation ?

pbludov commented:
> @pbludov , do you have time to finish this implementation ?

I'm on it

pbludov commented:
@Gaurav-Punjabi the work will continue in PR #9006
This PR is closed.
